### PR TITLE
Add Validate APIs for BackupLocation and ApplicationBackup.

### PR DESCRIPTION
**Pull request by Rohit-PX**
Why do we need this PR?
Stork/Torpedo and tests would require these PRs for application backup testing.

Signed-off-by: Rohit-PX <rohit@portworx.com>

### CODE REVIEW

File: k8s/k8s.go [hunk 1]
```diff
@@ -4487,6 +4491,29 @@ func (k *k8sOps) UpdateBackupLocation(backupLocation *v1alpha1.BackupLocation) (
...
+func (k *k8sOps) ValidateBackupLocation(name, namespace string, timeout, retryInterval time.Duration) error {
+	t := func() (interface{}, bool, error) {
+		if err := k.initK8sClient(); err != nil {
+			return "", true, err
+		}
+
+		resp, err := k.GetBackupLocation(name, namespace)
+		if err != nil {
+			return "", true, err
+		}
```
Comment (harsh-px) [inline k8s/k8s.go]: instead of just checking we can do a get, we should also validate the response as the APIs intention is validate.

Comment (disrani-px) [reply k8s/k8s.go]: There is nothing to validate, there is no status for the backupLocation

File: k8s/k8s.go [hunk 2]
```diff
@@ -4487,6 +4491,29 @@ func (k *k8sOps) UpdateBackupLocation(backupLocation *v1alpha1.BackupLocation) (
...
+		return "", true, &ErrFailedToValidateCustomSpec{
```
Comment (disrani-px) [inline k8s/k8s.go]: Need to return `"", false, nil` if err is nil. This will always return error right now.

File: k8s/k8s.go [hunk 3]
```diff
@@ -4533,6 +4560,34 @@ func (k *k8sOps) UpdateApplicationBackup(backup *v1alpha1.ApplicationBackup) (*v
 	return k.storkClient.Stork().ApplicationBackups(backup.Namespace).Update(backup)
 }
 
+func (k *k8sOps) ValidateApplicationBackup(name, namespace string, timeout, retryInterval time.Duration) error {
```
Comment (disrani-px) [inline k8s/k8s.go]: Also need to add `ValidateApplicationRestore`

File: k8s/k8s.go (hunk 1 above)
Comment (harsh-px) [reply k8s/k8s.go]: We can atleast check if `resp.Location` is populated.

Comment (disrani-px) [reply k8s/k8s.go]: This will just give false positives. Having anything set doesn't mean its valid.

File: k8s/k8s.go [hunk 4]
```diff
@@ -4533,6 +4564,32 @@ func (k *k8sOps) UpdateApplicationBackup(backup *v1alpha1.ApplicationBackup) (*v
...
+	t := func() (interface{}, bool, error) {
+		if err := k.initK8sClient(); err != nil {
+			return "", true, err
+		}
+
+		applicationbackup, err := k.GetApplicationBackup(name, namespace)
```
Comment (disrani-px) [inline k8s/k8s.go]: Need to check for err and retry

File: k8s/k8s.go [hunk 5]
```diff
@@ -4567,6 +4624,29 @@ func (k *k8sOps) DeleteApplicationRestore(name string, namespace string) error {
 	})
 }
 
+func (k *k8sOps) ValidateApplicationRestore(name, namespace string, timeout, retryInterval time.Duration) error {
+	t := func() (interface{}, bool, error) {
+		if err := k.initK8sClient(); err != nil {
+			return "", true, err
+		}
+
+		applicationrestore, err := k.storkClient.Stork().ApplicationRestores(namespace).Get(name, meta_v1.GetOptions{})
```
Comment (disrani-px) [inline k8s/k8s.go]: Need to check for err and retry

File: k8s/k8s.go (hunk 2 above)
Comment (Rohit-PX) [reply k8s/k8s.go]: Fixed.

File: k8s/k8s.go (hunk 3 above)
Comment (Rohit-PX) [reply k8s/k8s.go]: Fixed.

File: k8s/k8s.go (hunk 4 above)
Comment (Rohit-PX) [reply k8s/k8s.go]: Fixed.

File: k8s/k8s.go (hunk 5 above)
Comment (Rohit-PX) [reply k8s/k8s.go]: Fixed.

File: k8s/k8s.go (hunk 1 above)
Comment (Rohit-PX) [reply k8s/k8s.go]: Fixed.

File: k8s/k8s.go [hunk 6]
```diff
@@ -4567,6 +4625,32 @@ func (k *k8sOps) DeleteApplicationRestore(name string, namespace string) error {
...
+		if err != nil {
+			return "", true, nil
```
Comment (disrani-px) [inline k8s/k8s.go]: `return "", true, err`

File: k8s/k8s.go [hunk 7]
```diff
@@ -4487,6 +4493,29 @@ func (k *k8sOps) UpdateBackupLocation(backupLocation *v1alpha1.BackupLocation) (
...
+	t := func() (interface{}, bool, error) {
+		if err := k.initK8sClient(); err != nil {
+			return "", true, err
+		}
+
+		resp, err := k.GetBackupLocation(name, namespace)
+		if err != nil {
+			return "", true, &ErrFailedToValidateCustomSpec{
+				Name:  name,
+				Cause: fmt.Sprintf("BackupLocation %v: %v failed . Error: %v", name, namespace, err),
```
Comment (disrani-px) [inline k8s/k8s.go]: Name is already being passed into the error, don't need to print it again.

Comment (Rohit-PX) [reply k8s/k8s.go]: Fixed.

File: k8s/k8s.go (hunk 6 above)
Comment (Rohit-PX) [reply k8s/k8s.go]: Fixed.

**Pull request merged**